src/server/routes/contract/read/read-batch.ts

```diff
@@ -0,0 +1,170 @@
+import { standardResponseSchema } from "../../../schemas/sharedApiSchemas";
+import { getChainIdFromChain } from "../../../utils/chain";
+import { bigNumberReplacer } from "../../../utils/convertor";
+
+const MULTICALL3_ADDRESS = "0xcA11bde05977b3631167028862bE2a173976CA11";
+
+const MULTICALL3_AGGREGATE_ABI =
+  "function aggregate3((address target, bool allowFailure, bytes callData)[] calls) external payable returns ((bool success, bytes returnData)[])";
+
+const readCallRequestItemSchema = Type.Object({
+  contractAddress: Type.String(),
+  functionName: Type.String(),
+  functionAbi: Type.Optional(Type.String()),
+  args: Type.Optional(Type.Array(Type.Any())),
+});
+
+const readMulticallRequestSchema = Type.Object({
+  calls: Type.Array(readCallRequestItemSchema),
+  multicallAddress: Type.Optional(Type.String()),
+});
+
+const responseSchema = Type.Object({
+  results: Type.Array(
+    Type.Object({
+      success: Type.Boolean(),
+      result: Type.Any(),
+    }),
+  ),
+});
+
+const paramsSchema = Type.Object({
+  chain: Type.String(),
+});
+
+type RouteGeneric = {
+  Params: { chain: string };
+  Body: Static<typeof readMulticallRequestSchema>;
+  Reply: Static<typeof responseSchema>;
+};
+
+export async function readMulticall(fastify: FastifyInstance) {
+  fastify.route<RouteGeneric>({
+    method: "POST",
+    url: "/contract/:chain/read-batch",
+    schema: {
+      summary: "Batch read from multiple contracts",
+      description:
+        "Execute multiple contract read operations in a single call using Multicall3",
+      tags: ["Contract"],
+      operationId: "readMulticall",
+      params: paramsSchema,
+      body: readMulticallRequestSchema,
+      response: {
+        ...standardResponseSchema,
+        [StatusCodes.OK]: responseSchema,
+      },
+    },
+    handler: async (request, reply) => {
+      const { chain: chainSlug } = request.params;
+      const { calls, multicallAddress = MULTICALL3_ADDRESS } = request.body;
+
+      const chainId = await getChainIdFromChain(chainSlug);
+      const chain = await getChain(chainId);
+
+      try {
+        // Encode each read call
+        const encodedCalls = await Promise.all(
+          calls.map(async (call) => {
+            const contract = await getContract({
+              client: thirdwebClient,
+              chain,
+              address: call.contractAddress,
+            });
+
+            const method =
+              (call.functionAbi as unknown as AbiFunction) ??
+              (await resolveMethod(call.functionName)(contract));
+
+            const transaction = prepareContractCall({
+              contract,
+              method,
+              params: call.args || [],
+              // stubbing gas values so that the call can be encoded
+              maxFeePerGas: 30n,
+              maxPriorityFeePerGas: 1n,
+              value: 0n,
+            });
+
+            const calldata = await resolvePromisedValue(transaction.data);
```

Comment:
Don't use resolvePromisedValue. I should add a `@deprecated` tag on the function. Use this instead:
```suggestion
const callData = await encode(transaction);
```